examples/mahler_yum_2024/Mahler_Yum_2024.py

```diff
@@ -0,0 +1,544 @@
+    return jnp.logical_not(jnp.logical_and(period > retirement_age, working > 0))
+
+
+def savings_constraint(
+    net_income: FloatND, wealth: ContinuousState, saving: ContinuousAction
+) -> BoolND:
+    wealth = calc_savingsgrid(wealth)
+    saving = calc_savingsgrid(saving)
+    return net_income + (wealth) * r >= (saving)
+
+
+# ======================================================================================
+# Model specification and parameters
+# ======================================================================================
+
+
+ALIVE_REGIME = Regime(
+    name="alive",
+    utility=utility,
+    functions={
+        "disutil": disutil,
+        "fcost": fcost,
+        "cons_util": cons_util,
+        "cnow": cnow,
+        "income": income,
+        "benefits": benefits,
+        "adj_cost": adj_cost,
+        "net_income": net_income,
+        "taxed_income": taxed_income,
+        "pension": pension,
+        "retirement_constraint": retirement_constraint,
+        "savings_constraint": savings_constraint,
+    },
+    actions={
+        "working": DiscreteGrid(WorkingStatus),
+        "saving": LinspaceGrid(start=0, stop=49, n_points=50),
+        "effort": DiscreteGrid(Effort),
+    },
+    states={
+        "wealth": LinspaceGrid(start=0, stop=49, n_points=50),
+        "health": DiscreteGrid(HealthStatus),
+        "productivity_shock": DiscreteGrid(ProductivityShock),
+        "effort_t_1": DiscreteGrid(Effort),
+        "adjustment_cost": DiscreteGrid(AdjustmentCost),
+        "education": DiscreteGrid(EducationStatus),
+        "discount_factor": DiscreteGrid(DiscountFactor),
+        "productivity": DiscreteGrid(ProductivityType),
+        "health_type": DiscreteGrid(HealthType),
+    },
+    constraints={
+        "retirement_constraint": retirement_constraint,
+        "savings_constraint": savings_constraint,
+    },
+    transitions={
+        "alive": {
+            "next_wealth": next_wealth,
+            "next_health": next_health,
+            "next_productivity_shock": next_productivity_shock,
+            "next_discount_factor": next_discount_factor,
+            "next_adjustment_cost": next_adjustment_cost,
+            "next_effort_t_1": next_effort_t_1,
+            "next_health_type": next_health_type,
+            "next_education": next_education,
+            "next_productivity": next_productivity,
+        },
+        "dead": {"next_dead": lambda: Dead.dead},
+    },
+    regime_transition_probs=alive_to_dead,  # type: ignore[arg-type]
+)
+
+DEAD_REGIME = Regime(
+    name="dead",
+    utility=lambda dead: jnp.asarray([0.0]),  # noqa: ARG005
+    states={"dead": DiscreteGrid(Dead)},
+    actions={},
+    transitions={"dead": {"next_dead": lambda dead: Dead.dead}},  # noqa: ARG005
+    regime_transition_probs=lambda: {"alive": 0.0, "dead": 1.0},
+)
+
+MAHLER_YUM_MODEL = Model([ALIVE_REGIME, DEAD_REGIME], n_periods=n)
+
+
+########################
```

Comment:
Not crucial for now, but I'd really like us to use the possibilities of pytrees here. This would be much clearer if it were in a hierarchical structure, like:

```py
params = {
    nu: {
        ad: 0.8,
        h: jnp.array([2.63, 1.66, 1.27, 1.71]),
        u: jnp.array([2.41, 1.18, 1.39, 2.41]),
    },
    ...
}
```